### PR TITLE
[WIP] OpenShift Support Added.

### PR DESCRIPTION
**Pull request by sjug**
For review and comments.

Missing resource update changes.

Do not merge.

### CODE REVIEW

**Comment by sjug** [comment]
@adrianchiris PTAL

**Comment by adrianchiris** [comment]
@sjug could you rebase the PR so we can start review,

File: example/deploy/role.yaml [hunk 1]
```diff
@@ -171,31 +171,76 @@ metadata:
...
-  - watch
-- apiGroups:
+  - apiGroups:
+    - rbac.authorization.k8s.io
+    resources:
+    - roles
+    - rolebindings
+    verbs:
+    - '*'
+  - apiGroups:
```
Comment (adrianchiris) [inline example/deploy/role.yaml:181]: id like to keep cluster scope resource access to a minimum as well as the verbs the operator is allowed to perform on them.

can you give a bit more context why these new additions are needed ?

File: example/deploy/role.yaml [hunk 2]
```diff
@@ -171,31 +171,76 @@ metadata:
...
+    - patch
+    - update
     - watch
-- apiGroups:
+  - apiGroups:
     - apps
-  resources:
-    - daemonsets
-  verbs:
+    resources:
```
Comment (adrianchiris) [inline example/deploy/role.yaml:203]: do we need to be able to CRUD all resources under apps ?
can this be limited to a subset ?

File: manifests/stage-ofed-driver/0100_service_account.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+kind: ServiceAccount
```
Comment (adrianchiris) [inline manifests/stage-ofed-driver/0100_service_account.yaml:2]: can you provide some info about why is that needed ?

File: manifests/stage-ofed-driver/0100_service_account.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+metadata:
+  name: ofed-driver
+  namespace: mlnx-network-operator
```
Comment (adrianchiris) [inline manifests/stage-ofed-driver/0100_service_account.yaml:5]: as an FYI, the resources network-operator deploys are now in a fixed namespace : `mlnx-network-operator-resources`
does this service account need be deployed in that namespace ? if so, you should use the template : `namespace: {{ .RuntimeSpec.Namespace }}`

File: manifests/stage-ofed-driver/0100_service_account.yaml (hunk 3 above)
Comment (adrianchiris) [reply manifests/stage-ofed-driver/0100_service_account.yaml:2]: also is it needed per stage ? i see you added it just for ofed-driver stage.

File: pkg/state/state_skel.go [hunk 5]
```diff
@@ -91,32 +92,103 @@ func (s *stateSkel) updateObj(obj *unstructured.Unstructured) error {
...
+		kind == "PrometheusRule" {
+		return true
+	}
+	return false
+}
+
+func checkNestedFields(found bool, err error) {
+	if !found || err != nil {
+		log.Error(err, "")
+		os.Exit(1)
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:112]: why not just return an error ?
os.Exit() will exit the entire process right ?

File: pkg/state/state_skel.go [hunk 6]
```diff
@@ -91,32 +92,103 @@ func (s *stateSkel) updateObj(obj *unstructured.Unstructured) error {
...
+	}
+}
+
+func updateResourceVersion(req, found *unstructured.Unstructured) error {
+	kind := found.GetKind()
+
+	if needToUpdateResourceVersion(kind) {
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:119]: can you add a comment on why we need to update resource version ?

also i see you are just updating the field with the same value, is that ok ?

im probably lacking some k8s specific knowledge about `resourceVersion` so apologies if the questions seem obvious

File: pkg/state/state_skel.go [hunk 7]
```diff
@@ -91,32 +92,103 @@ func (s *stateSkel) updateObj(obj *unstructured.Unstructured) error {
...
+		version, fnd, err := unstructured.NestedString(found.Object, "metadata", "resourceVersion")
+		checkNestedFields(fnd, err)
+
+		if err := unstructured.SetNestedField(req.Object, version, "metadata", "resourceVersion"); err != nil {
+			return errors.Wrap(err, "Couldn't update ResourceVersion")
+		}
+	}
+
+	if kind == "Service" {
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:128]: can you add a comment on why we need to update ClusterIP ?

also i see you are just updating the field with the same value, is that ok ?

im probably lacking some k8s specific knowledge about `ClusterIP ` so apologies if the questions seem obvious

File: pkg/state/state_skel.go [hunk 8]
```diff
@@ -91,32 +92,103 @@ func (s *stateSkel) updateObj(obj *unstructured.Unstructured) error {
...
 		}
-		if !k8serrors.IsAlreadyExists(err) {
-			// Some error occurred
-			return err
+
+		if err != nil {
+			return errors.Wrap(err, "unexpected error")
+		}
+
+		// Short circuit updating ServiceAccounts and Pods
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:172]: is that because updates are not allowed on them ?
can you add the reason in the comment ?

**Review by adrianchiris — COMMENTED** [review]
Thanks for proposing the PR !

comments inline

my main concerns are:
1. the additional permissions added to the operator at cluster scope
2. the addition of role/role binding/service account in stage-ofed

so more info on those is appreciated also i see the additions of objects was only done in stage-ofed, is it not needed for other stages ? (nv peer mem or rdma device plugin ?)

Also id like to flag the openshift specific resources they may not be deployed in non openshift scenarios when helm support is added.

File: example/deploy/role.yaml (hunk 1 above)
Comment (sjug) [reply example/deploy/role.yaml:181]: @adrianchiris Are you asking specifically about this `"" apiGroup` or in general about all the cluster scoped resources?

File: example/deploy/role.yaml (hunk 2 above)
Comment (sjug) [reply example/deploy/role.yaml:203]: As a first pass I likely added more operations under apps in the interest of expedience, I can narrow them down.

File: manifests/stage-ofed-driver/0100_service_account.yaml (hunk 3 above)
Comment (sjug) [reply manifests/stage-ofed-driver/0100_service_account.yaml:2]: In order to add the SecurityContextConstraints (SCC) below, we need the corresponding ServiceAccount, etc.

File: manifests/stage-ofed-driver/0100_service_account.yaml (hunk 4 above)
Comment (sjug) [reply manifests/stage-ofed-driver/0100_service_account.yaml:5]: Ack changed the namespaces for the `manifests/stage-ofed-driver/*.yaml`

File: pkg/state/state_skel.go (hunk 5 above)
Comment (sjug) [reply pkg/state/state_skel.go:112]: Ack changed.

File: pkg/state/state_skel.go (hunk 6 above)
Comment (sjug) [reply pkg/state/state_skel.go:119]: That's just the standard pattern for k8s afaik, but some OpenShift resources don't need it/are different API wise.

File: pkg/state/state_skel.go (hunk 7 above)
Comment (sjug) [reply pkg/state/state_skel.go:128]: Just lessons learned from other operator issues, I will check into why this is required for `Services`.

File: pkg/state/state_skel.go (hunk 8 above)
Comment (sjug) [reply pkg/state/state_skel.go:172]: Added comment

File: example/deploy/role.yaml (hunk 1 above)
Comment (adrianchiris) [reply example/deploy/role.yaml:181]: in general, sorry i should have placed this comment on top at L#174

File: manifests/stage-ofed-driver/0300_rolebinding.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: ofed-driver
+  namespace: {{ .RuntimeSpec.Namespace }}
+subjects:
+- kind: ServiceAccount
+  name: ofed-driver
+  namespace: {{ .RuntimeSpec.Namespace }}
+userNames:
```
Comment (adrianchiris) [inline manifests/stage-ofed-driver/0300_rolebinding.yaml:15]: is the `userNames` attribute intended for older versions of K8s ?  i see the V1 api does not have this attribute[1]

[1] https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#rolebinding-v1-rbac-authorization-k8s-io

File: manifests/stage-ofed-driver/0400_scc.openshift.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
+# Please edit the object below. Lines beginning with a '#' will be ignored,
```
Comment (adrianchiris) [inline manifests/stage-ofed-driver/0400_scc.openshift.yaml:1]: I believe this will also be needed for nv-peer-mem driver as well,
IMO we should have a separate stage to deploy OpenShift's security related orbjects. WDYT ?

the alternative is having this duplicated in every stage as `stage-ofed-driver` can be opted out.

File: pkg/state/state_skel.go [hunk 11]
```diff
@@ -91,32 +91,104 @@ func (s *stateSkel) updateObj(obj *unstructured.Unstructured) error {
...
+		return errors.Wrap(err, "nested field error")
+	}
+}
+
+func updateResourceVersion(req, found *unstructured.Unstructured) error {
+	kind := found.GetKind()
+
+	if needToUpdateResourceVersion(kind) {
+		version, fnd, err := unstructured.NestedString(found.Object, "metadata", "resourceVersion")
+		checkNestedFields(fnd, err)
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:119]: `checkNestedFields` will now silently fail, i don't think we want that right ?

As we (now) don't call `os.exit` i think an if statement is needed to check for errors.
either remove `checkNestedFields` and implement the logic directly or have it return an error and check it here.

File: pkg/state/state_skel.go [hunk 12]
```diff
@@ -91,32 +91,104 @@ func (s *stateSkel) updateObj(obj *unstructured.Unstructured) error {
...
+	return false
+}
+
+func checkNestedFields(found bool, err error) {
+	if !found || err != nil {
+		return errors.Wrap(err, "nested field error")
+	}
+}
+
+func updateResourceVersion(req, found *unstructured.Unstructured) error {
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:114]: we update both resourceVersion and clusterIP for service.
down the road it might update some other fields depending on the object kind.

perhaps call it: `updateInternalFields` ?

File: pkg/state/state_skel.go (hunk 6 above)
Comment (adrianchiris) [reply pkg/state/state_skel.go:118]: From what i can tell from the below this is needed for implementing optimistic concurrency: 

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency

File: pkg/state/state_skel.go [hunk 13]
```diff
@@ -91,32 +91,109 @@ func (s *stateSkel) updateObj(obj *unstructured.Unstructured) error {
...
+	kind := found.GetKind()
+
+	if needToUpdateResourceVersion(kind) {
+		version, fnd, err := unstructured.NestedString(found.Object, "metadata", "resourceVersion")
+		if err := checkNestedFields(fnd, err); err != nil {
+			return errors.Wrap(err, "Couldn't check nested fields.")
+		}
+
+		if err := unstructured.SetNestedField(req.Object, version, "metadata", "resourceVersion"); err != nil {
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:124]: nit: why not use:  `req.SetResourceVersion(version)` ?
